File: C/integer_triangle.c
#include <stdio.h>

int	max(int a, int b)
{
	if (a > b)
		return (a);
	return (b);
}

int	main(void)
{
	int	n;
	scanf("%d", &n);

	int	tri[n][n];
	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j <= i; j++)
			scanf("%d", &tri[i][j]);
	}

	int	dp[n][n];

	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < n; j++)
			dp[i][j] = 0;
	}
	dp[0][0] = tri[0][0];
	dp[1][0] = dp[0][0] + tri[1][0];
	dp[1][1] = dp[0][0] + tri[1][1];

	for (int i = 2; i < n; i++)
	{
		for (int j = 0; j <= i; j++)
		{
			if (j == 0)
				dp[i][j] = dp[i-1][j] + tri[i][j];
			else if (j == i)
				dp[i][j] = dp[i-1][j-1] + tri[i][j];
			else
				dp[i][j] = max(dp[i-1][j-1] + tri[i][j], dp[i-1][j] + tri[i][j]);
		}
	}

	int	dst = dp[n-1][0];

	for (int i = 1; i < n; i++)
	{
		if (dst < dp[n-1][i])
			dst = dp[n-1][i];
	}
	printf("%d", dst);
	return (0);
}

File: JAVA/src/IntegerTriangle.java
import java.util.Scanner;

public class IntegerTriangle {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[][] tri = new int[n][n];
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j <= i; ++j)
                tri[i][j] = sc.nextInt();
        }

        for (int i = n-1; i > 0; --i) {
            for (int j = 0; j < i; ++j)
                tri[i-1][j] += Math.max(tri[i][j], tri[i][j+1]);
        }

        System.out.println(tri[0][0]);
    }
}
